apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/MemoryCache.kt

```diff
@@ -0,0 +1,99 @@
+package com.apollographql.apollo.cache.normalized
+
+import com.apollographql.apollo.cache.ApolloCacheHeaders
+import com.apollographql.apollo.cache.CacheHeaders
+import com.apollographql.apollo.cache.normalized.internal.LruCache
+import com.apollographql.apollo.cache.normalized.internal.Platform
+import okio.internal.commonAsUtf8ToByteArray
+import kotlin.reflect.KClass
+
+class MemoryCache(
+    private val maxSizeBytes: Int,
+    private val expireAfterMillis: Long = -1,
+) : NormalizedCache() {
+  private val lruCache = LruCache<String, CacheEntry>(maxSize = maxSizeBytes) { key, cacheEntry ->
+    key.commonAsUtf8ToByteArray().size + (cacheEntry?.sizeInBytes ?: 0)
+  }
+
+  val size: Int
+    get() = lruCache.size()
+
+  override fun loadRecord(key: String, cacheHeaders: CacheHeaders): Record? {
+    val cachedEntry = lruCache[key]
+    return if (cachedEntry == null || cachedEntry.isExpired) {
+      if (cachedEntry != null) {
+        lruCache.remove(key)
+      }
+      nextCache?.loadRecord(key, cacheHeaders)
+    } else {
+      if (cacheHeaders.hasHeader(ApolloCacheHeaders.EVICT_AFTER_READ)) {
+        lruCache.remove(key)
+      }
+      cachedEntry.record
+    }
+  }
+
+  override fun clearAll() {
+    lruCache.clear()
+    nextCache?.clearAll()
+  }
+
+  override fun remove(cacheKey: CacheKey, cascade: Boolean): Boolean {
```

Comment:
Not specifically linked to this PR but I think the `cascade` parameter should be removed from the `NormalizedCache` API and handled in a more global way. That'll make it easier to implement caches and avoid behavioral differences between caches.

Reply:
Then we need to change the API as remove now needs to return store record instead of boolean.
Don't quite remember if there was particular reason of doing this on the low level end, we might have been thinking if we can run cascade deleting in a batch, we have to think about this carefully. Also current behaviour that we remove everything from current cache and only after that move to next chained one. With proposed changes we change that to remove one item on each cache before moving to nested referenced items

Reply:
Yup, agreed